text/0145-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,718 @@
+* Function signatures introduced by [PPP#7](https://github.com/w3f/PPPs/pull/7) have been used in this RFC, as the PPP has already been [properly implemented](https://github.com/paritytech/substrate/pull/11490) and [documented](https://github.com/w3f/polkadot-spec/pull/592/files). However, it has never been officially adopted, nor have its functions been in use.
+* Return values were harmonized to `i64` everywhere where they represent either a positive outcome as a positive integer or a negative outcome as a negative error code.
+* `ext_offchain_network_peer_id_version_1` now returns a result code instead of silently failing if the network status is unavailable.
+* Added new versions of `ext_misc_runtime_version` and `ext_offchain_random_seed`.
+* Addressed discussions from the original RFC-4 discussion thread.
+
+## Motivation
+
+The heap allocation of the runtime is currently controlled by the host using a memory allocator on the host side.
+
+The API of many host functions contains buffer allocations. For example, when calling `ext_hashing_twox_256_version_1`, the host allocates a 32-byte buffer using the host allocator, and returns a pointer to this buffer to the runtime. The runtime later has to call `ext_allocator_free_version_1` on this pointer to free the buffer.
+
+Even though no benchmark has been done, it is pretty obvious that this design is very inefficient. To continue with the example of `ext_hashing_twox_256_version_1`, it would be more efficient to instead write the output hash to a buffer allocated by the runtime on its stack and passed by pointer to the function. Allocating a buffer on the stack, in the worst case, consists simply of decreasing a number; in the best case, it is free. Doing so would save many VM memory reads and writes by the allocator, and would save a function call to `ext_allocator_free_version_1`.
+
+Furthermore, the existence of the host-side allocator has become questionable over time. It is implemented in a very naive way, and for determinism and backwards compatibility reasons, it needs to be implemented exactly identically in every client implementation. Runtimes make substantial use of heap memory allocations, and each allocation needs to go through the runtime <-> host boundary twice (once for allocating and once for freeing). Moving the allocator to the runtime side would be a good idea, although it would increase the runtime size. But before the host-side allocator can be deprecated, all the host functions that use it must be updated to avoid using it.
+
+## Stakeholders
+
+No attempt was made to convince stakeholders.
+
+## Explanation
+
+### New definitions
+
+#### <a name="new-def-i"></a>New Definition I: Runtime Optional Positive Integer
+
+The Runtime optional positive integer is a signed 64-bit value. Positive values in the range of [0..2³²) represent corresponding unsigned 32-bit values. The value of `-1` represents a non-existing value (an _absent_ value). All other values are invalid.
```

Comment:
1. In this definition "Runtime" is capitalized, and in the one below it's lowercase. (:
2. What does "All other values are invalid" entail? The runtime can pass these to the host so we need to define *exactly* what happens if such a value is passed. But maybe a better idea is to just *not* have any invalid values? If we need a *signed* integer then this we could make it so that every value is valid except `i64::MIN`/`0x8000000000000000` which is special-cased to mean nothing (i.e. it's `None` when only the most significant bit is set; this gives the full symmetric `-i64::MAX..i64::MAX` range available to the users). If we need an *unsigned* integer then we could make every value valid except `u64::MAX`/`0xffffffffffffffff`.